src/four_sect_dict.rs

```diff
@@ -157,6 +174,128 @@ impl FourSectDict {
         })
     }
 
+    /// read the whole dictionary section including control information
+    pub fn read_nt<R: BufRead>(reader: &mut R, opts: Options) -> Result<(Self, Vec<TripleId>), DictReadError> {
+        // TODO switch to sophia
+        let quads = RdfParser::from_format(NTriples).for_reader(reader);
+        let timer = std::time::Instant::now();
+        let mut raw_triples = Vec::new(); // Store raw triples
+
+        // TODO: compare times with Vec followed by parallel sort vs times with BTreeSet
+        let mut subject_terms = BTreeSet::new();
+        let mut object_terms = BTreeSet::new();
+        let mut predicate_terms = BTreeSet::new();
+
+        for q in quads {
+            let q = match q {
+                Ok(v) => v,
+                Err(e) => return Err(DictReadError::Other(format!("error parsing triple file: {e}"))),
+            }; //propagate the error  
+
+            let subj_str = term_to_hdt_bgp_str(&q.subject.into());
+            let pred_str = term_to_hdt_bgp_str(&q.predicate.into());
+            let obj_str = term_to_hdt_bgp_str(&q.object);
+
+            subject_terms.insert(subj_str.clone());
+            predicate_terms.insert(pred_str.clone());
+            object_terms.insert(obj_str.clone());
+
+            raw_triples.push((subj_str, pred_str, obj_str)); // Store for later encoding
+        }
+        if predicate_terms.is_empty() {
+            warn!("no triples found in provided RDF");
+        }
+        raw_triples.sort_unstable(); // Faster than stable sort
+        raw_triples.dedup();
+
+        let mut shared_terms = subject_terms.clone();
+        shared_terms.retain(|item| object_terms.contains(item));
+        let unique_subject_terms: BTreeSet<String> = subject_terms.difference(&shared_terms).cloned().collect();
+        let unique_object_terms: BTreeSet<String> = object_terms.difference(&shared_terms).cloned().collect();
+
+        // let mut so_id_map: HashMap<String, u32> = HashMap::new();
+        // let mut pred_id_map: HashMap<String, u32> = HashMap::new();
+        // let mut subject_id_map: HashMap<String, u32> = HashMap::new();
+        // let mut object_id_map: HashMap<String, u32> = HashMap::new();
+
+        /*
+        https://www.w3.org/submissions/2011/SUBM-HDT-20110330/#dictionaryEncoding
+        Four subsets are mapped as follows (for a graph G with SG, PG, OG the different subjects, predicates and objects):
+        1. Common subject-objects (SOG) with IDs from 1 to |SOG|
+        2. Non common subjects (SG-SOG), mapped to [|SOG| +1, |SG|]
+        3. Non common objects (OG-SOG), in [|SOG|+1, |OG|]
+        4. Predicates, mapped to [1, |PG|].
+         */
+
```

Comment:
This might be a memory vs time trade off, do we build the `Id` encoded triples using these HashMaps (faster?), or rely on the dictionary `string_to_id()` to encode the triples (less memory usage?)